File: packages/jetbrains/src/main/kotlin/com/wave/jetbrains/stdio/StdioAgent.kt
package com.wave.jetbrains.stdio

import kotlinx.coroutines.CompletableDeferred
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

/**
 * Callbacks bridging stdio notifications to the host/webview layer.
 * Each corresponds to a server→client notification.
 */
interface AgentCallbacks {
    fun onMessagesChange(messages: JsonElement?) {}
    fun onUserMessageAdded(message: JsonElement?) {}
    fun onAssistantMessageAdded(message: JsonElement?) {}
    fun onAssistantContentUpdated(messageId: String, accumulated: String, stage: String) {}
    fun onAssistantReasoningUpdated(messageId: String, accumulated: String, stage: String) {}
    fun onToolBlockUpdated(params: JsonElement?) {}
    fun onErrorBlockAdded(error: String) {}
    fun onCompactBlockAdded(content: String) {}
    fun onLoadingChange(loading: Boolean) {}
    fun onCommandRunningChange(running: Boolean) {}
    fun onQueuedMessagesChange(messages: JsonElement?) {}
    fun onTasksChange(tasks: JsonElement?) {}
    fun onSessionIdChange(sessionId: String) {}
    fun onPermissionModeChange(mode: String) {}
    fun onMcpServersChange(servers: JsonElement?) {}
    fun onWorkdirChange(workdir: String) {}
    fun onPermissionRequest(requestId: String, context: JsonElement?) {}
    fun onBangMessageAdded() {}
    fun onBangMessageUpdated() {}
    fun onBangMessageCompleted() {}
    fun onNotificationMessageAdded(message: JsonObject) {}
    fun onError(message: String) {}
}

/**
 * Typed business wrapper over StdioClient. Caches state and routes notifications.
 * Mirrors packages/vsce/src/stdio/stdioAgent.ts.
 */
class StdioAgent(
    private val client: StdioClient,
    private val router: NotificationRouter,
    private val callbacks: AgentCallbacks,
) {
    @Volatile var sessionId: String? = null
        private set
    @Volatile var workingDirectory: String? = null
        private set
    @Volatile var latestTotalTokens: Int = 0
        private set
    @Volatile var permissionMode: String? = null
        private set
    @Volatile var serverVersion: String? = null
        private set

    fun handleNotification(method: String, params: JsonElement?) {
        when (method) {
            "messagesChange" -> callbacks.onMessagesChange(params?.jsonObject?.get("messages"))
            "userMessageAdded" -> callbacks.onUserMessageAdded(params?.jsonObject?.get("message"))
            "assistantMessageAdded" -> callbacks.onAssistantMessageAdded(params?.jsonObject?.get("message"))
            "assistantContentUpdated" -> {
                val o = params?.jsonObject
                callbacks.onAssistantContentUpdated(
                    o?.get("messageId")?.jsonPrimitive?.content ?: "",
                    o?.get("accumulated")?.jsonPrimitive?.content ?: "",
                    o?.get("stage")?.jsonPrimitive?.content ?: "",
                )
            }
            "assistantReasoningUpdated" -> {
                val o = params?.jsonObject
                callbacks.onAssistantReasoningUpdated(
                    o?.get("messageId")?.jsonPrimitive?.content ?: "",
                    o?.get("accumulated")?.jsonPrimitive?.content ?: "",
                    o?.get("stage")?.jsonPrimitive?.content ?: "",
                )
            }
            "toolBlockUpdated" -> callbacks.onToolBlockUpdated(params)
            "errorBlockAdded" -> callbacks.onErrorBlockAdded(params?.jsonObject?.get("error")?.jsonPrimitive?.content ?: "")
            "loadingChange" -> {
                val o = params?.jsonObject
                o?.get("latestTotalTokens")?.jsonPrimitive?.intOrNull?.let { latestTotalTokens = it }
                callbacks.onLoadingChange(o?.get("loading")?.jsonPrimitive?.content?.toBoolean() ?: false)
            }
            "commandRunningChange" -> callbacks.onCommandRunningChange(
                params?.jsonObject?.get("running")?.jsonPrimitive?.content?.toBoolean() ?: false
            )
            "queuedMessagesChange" -> callbacks.onQueuedMessagesChange(params?.jsonObject?.get("messages"))
            "tasksChange" -> callbacks.onTasksChange(params?.jsonObject?.get("tasks"))
            "sessionIdChange" -> {
                val id = params?.jsonObject?.get("sessionId")?.jsonPrimitive?.content ?: ""
                sessionId = id
                callbacks.onSessionIdChange(id)
            }
            "permissionModeChange" -> {
                val mode = params?.jsonObject?.get("mode")?.jsonPrimitive?.content ?: ""
                permissionMode = mode
                callbacks.onPermissionModeChange(mode)
            }
            "mcpServersChange" -> callbacks.onMcpServersChange(params?.jsonObject?.get("servers"))
            "workdirChange" -> {
                val workdir = params?.jsonObject?.get("workdir")?.jsonPrimitive?.content ?: ""
                workingDirectory = workdir
                callbacks.onWorkdirChange(workdir)
            }
            "permissionRequest" -> {
                val o = params?.jsonObject
                callbacks.onPermissionRequest(
                    o?.get("requestId")?.jsonPrimitive?.content ?: "",
                    o?.get("context"),
                )
            }
            "bangMessageAdded" -> callbacks.onBangMessageAdded()
            "bangMessageUpdated" -> callbacks.onBangMessageUpdated()
            "bangMessageCompleted" -> callbacks.onBangMessageCompleted()
            "notificationMessageAdded" -> callbacks.onNotificationMessageAdded(params?.jsonObject ?: JsonObject(emptyMap()))
            "compactBlockAdded" -> callbacks.onCompactBlockAdded(params?.jsonObject?.get("content")?.jsonPrimitive?.content ?: "")
            else -> {}
        }
    }

    // ── RPC requests ──────────────────────────────────────────────

    suspend fun initialize(params: JsonObject): InitializeResult {
        val res = client.request("initialize", params)?.jsonObject
            ?: throw StdioClientException("initialize returned null")
        sessionId = res["sessionId"]?.jsonPrimitive?.content
        workingDirectory = res["workingDirectory"]?.jsonPrimitive?.content
        permissionMode = res["permissionMode"]?.jsonPrimitive?.content
        res["latestTotalTokens"]?.jsonPrimitive?.intOrNull?.let { latestTotalTokens = it }
        res["serverVersion"]?.jsonPrimitive?.content?.let { serverVersion = it }
        sessionId?.let { router.register(it, this) }
        return InitializeResult(
            sessionId = sessionId,
            workingDirectory = workingDirectory,
            permissionMode = permissionMode,
            latestTotalTokens = latestTotalTokens,
            serverVersion = serverVersion,
        )
    }

    suspend fun destroy() {
        sessionId?.let { router.unregister(it) }
        runCatching { client.request("destroy", sessionId = sessionId) }
    }

    suspend fun sendMessage(text: String, images: JsonElement? = null, force: Boolean = false) {
        client.request("sendMessage", buildJsonObject {
            put("text", text)
            if (images != null) put("images", images)
            if (force) put("force", true)
        }, sessionId)
    }

    suspend fun bang(command: String) {
        client.request("bang", buildJsonObject { put("command", command) }, sessionId)
    }

    suspend fun abortMessage() { client.request("abortMessage", sessionId = sessionId) }
    suspend fun clearMessages() { client.request("clearMessages", sessionId = sessionId) }

    suspend fun restoreSession(sessionId: String) {
        client.request("restoreSession", buildJsonObject { put("sessionId", sessionId) }, this.sessionId)
    }

    suspend fun setPermissionMode(mode: String) {
        client.request("setPermissionMode", buildJsonObject { put("mode", mode) }, sessionId)
    }

    suspend fun deleteQueuedMessage(index: Int) {
        client.request("deleteQueuedMessage", buildJsonObject { put("index", index) }, sessionId)
    }

    suspend fun updateQueuedMessage(id: String, text: String, images: JsonElement? = null): Boolean {
        val result = client.request("updateQueuedMessage", buildJsonObject {
            put("id", id)
            put("text", text)
            if (images != null) put("images", images)
        }, sessionId)
        return result?.jsonObject?.get("ok")?.jsonPrimitive?.booleanOrNull ?: false
    }

    suspend fun deleteQueuedMessageById(id: String) {
        client.request("deleteQueuedMessageById", buildJsonObject { put("id", id) }, sessionId)
    }

    suspend fun getSlashCommands(): JsonElement? = client.request("getSlashCommands", sessionId = sessionId)

    suspend fun updateConfig(params: JsonObject) {
        val oldSessionId = sessionId
        val result = client.request("updateConfig", params, sessionId)?.jsonObject
        val newSessionId = result?.get("sessionId")?.jsonPrimitive?.content
        if (oldSessionId != null && newSessionId != null && newSessionId != oldSessionId) {
            router.unregister(oldSessionId)
            sessionId = newSessionId
            router.register(newSessionId, this)
        }
    }

    suspend fun rewindToMessage(messageId: String): String {
        val result = client.request("rewindToMessage", buildJsonObject { put("messageId", messageId) }, sessionId)
        return result?.jsonObject?.get("inputContent")?.jsonPrimitive?.content ?: ""
    }

    suspend fun getMcpServers(): JsonElement =
        client.request("getMcpServers", sessionId = sessionId) ?: JsonObject(emptyMap())

    suspend fun connectMcpServer(serverName: String): Boolean {
        val result = client.request("connectMcpServer", buildJsonObject { put("serverName", serverName) }, sessionId)
        return result?.jsonObject?.get("success")?.jsonPrimitive?.booleanOrNull ?: false
    }

    suspend fun disconnectMcpServer(serverName: String): Boolean {
        val result = client.request("disconnectMcpServer", buildJsonObject { put("serverName", serverName) }, sessionId)
        return result?.jsonObject?.get("success")?.jsonPrimitive?.booleanOrNull ?: false
    }

    // ── RPC notification (fire-and-forget) ────────────────────────

    suspend fun sendPermissionResponse(requestId: String, decision: JsonObject) {
        client.notify("permissionResponse", buildJsonObject {
            put("requestId", requestId)
            put("decision", decision)
        }, sessionId)
    }

    // ── Utility RPCs (standalone, mirror VSCE services) ───────────

    suspend fun searchFiles(query: String, workdir: String, maxResults: Int = 20): JsonElement =
        client.request("searchFiles", buildJsonObject {
            put("query", query)
            put("maxResults", maxResults)
            put("workdir", workdir)
        }) ?: JsonObject(emptyMap())

    suspend fun listSessions(workdir: String): JsonElement =
        client.request("listSessions", buildJsonObject { put("workdir", workdir) }) ?: JsonObject(emptyMap())

    suspend fun getPromptHistory(): JsonElement =
        client.request("getPromptHistory") ?: JsonObject(emptyMap())

    suspend fun searchPromptHistory(query: String): JsonElement =
        client.request("searchPromptHistory", buildJsonObject { put("query", query) }) ?: JsonObject(emptyMap())

    suspend fun listPlugins(workdir: String): JsonElement =
        client.request("listPlugins", buildJsonObject { put("workdir", workdir) }) ?: JsonObject(emptyMap())

    suspend fun installPlugin(pluginId: String, workdir: String, scope: String? = null): JsonElement =
        client.request("installPlugin", buildJsonObject {
            put("pluginId", pluginId)
            if (scope != null) put("scope", scope)
            put("workdir", workdir)
        }) ?: JsonObject(emptyMap())

    suspend fun uninstallPlugin(pluginId: String, workdir: String): JsonElement =
        client.request("uninstallPlugin", buildJsonObject {
            put("pluginId", pluginId)
            put("workdir", workdir)
        }) ?: JsonObject(emptyMap())

    suspend fun enablePlugin(pluginId: String, workdir: String, scope: String? = null): JsonElement =
        client.request("enablePlugin", buildJsonObject {
            put("pluginId", pluginId)
            if (scope != null) put("scope", scope)
            put("workdir", workdir)
        }) ?: JsonObject(emptyMap())

    suspend fun disablePlugin(pluginId: String, workdir: String, scope: String? = null): JsonElement =
        client.request("disablePlugin", buildJsonObject {
            put("pluginId", pluginId)
            if (scope != null) put("scope", scope)
            put("workdir", workdir)
        }) ?: JsonObject(emptyMap())

    suspend fun updatePlugin(pluginId: String, workdir: String): JsonElement =
        client.request("updatePlugin", buildJsonObject {
            put("pluginId", pluginId)
            put("workdir", workdir)
        }) ?: JsonObject(emptyMap())

    suspend fun listMarketplaces(workdir: String): JsonElement =
        client.request("listMarketplaces", buildJsonObject { put("workdir", workdir) }) ?: JsonObject(emptyMap())

    suspend fun addMarketplace(input: String, workdir: String): JsonElement =
        client.request("addMarketplace", buildJsonObject {
            put("input", input)
            put("workdir", workdir)
        }) ?: JsonObject(emptyMap())

    suspend fun removeMarketplace(name: String, workdir: String): JsonElement =
        client.request("removeMarketplace", buildJsonObject {
            put("name", name)
            put("workdir", workdir)
        }) ?: JsonObject(emptyMap())

    suspend fun updateMarketplace(workdir: String, name: String? = null): JsonElement =
        client.request("updateMarketplace", buildJsonObject {
            if (name != null) put("name", name)
            put("workdir", workdir)
        }) ?: JsonObject(emptyMap())

    suspend fun getAuthStatus(): JsonElement =
        client.request("getAuthStatus") ?: JsonObject(emptyMap())

    suspend fun login(): JsonElement =
        client.request("login") ?: JsonObject(emptyMap())

    suspend fun logout(): JsonElement =
        client.request("logout") ?: JsonObject(emptyMap())
}

data class InitializeResult(
    val sessionId: String?,
    val workingDirectory: String?,
    val permissionMode: String?,
    val latestTotalTokens: Int,
    val serverVersion: String? = null,
)

private val JsonPrimitive.intOrNull: Int? get() = content.toIntOrNull()

File: packages/vsce/src/stdio/stdioAgent.ts
/**
 * StdioAgent — typed wrapper around StdioClient that mirrors the Agent API.
 *
 * In the single-shared-process architecture, all sessions share one
 * StdioClient. The NotificationRouter dispatches incoming notifications to
 * the appropriate StdioAgent via `handleNotification()`. The agent no longer
 * subscribes directly to the client.
 *
 * All session-scoped requests carry `this.sessionId` on the JSON-RPC envelope
 * so the server can route them to the right Agent.
 *
 * `destroy()` only unregisters from the router and sends the destroy request;
 * it does NOT dispose the shared StdioClient.
 */

import type {
    Message,
    Task,
    QueuedMessage,
    McpServerStatus,
    PermissionMode,
    PermissionDecision,
    ToolPermissionContext,
    ToolBlockUpdateCallbackParams,
    SlashCommand,
    McpServerConfig,
} from 'wave-agent-sdk';
import { StdioClient } from './stdioClient';
import { NotificationRouter } from './notificationRouter';

// ── Params / Results ─────────────────────────────────────────────

export interface InitializeParams {
    workdir?: string;
    restoreSessionId?: string;
    apiKey?: string;
    baseURL?: string;
    serverUrl?: string;
    defaultHeaders?: Record<string, string>;
    model?: string;
    fastModel?: string;
    language?: string;
    permissionMode?: PermissionMode;
    tools?: string[];
    allowedTools?: string[];
    disallowedTools?: string[];
    pluginDirs?: string[];
    mcpServers?: Record<string, McpServerConfig>;
    clientVersion?: string;
}

export interface InitializeResult {
    sessionId: string;
    workingDirectory: string;
    permissionMode: PermissionMode;
    latestTotalTokens: number;
    serverVersion?: string;
}

export interface UpdateConfigParams {
    apiKey?: string;
    baseURL?: string;
    serverUrl?: string;
    defaultHeaders?: Record<string, string>;
    model?: string;
    fastModel?: string;
    language?: string;
}

// ── Callbacks (mirror AgentCallbacks) ────────────────────────────

export interface StdioAgentCallbacks {
    onMessagesChange?: (messages: Message[]) => void;
    onUserMessageAdded?: (message: Message) => void;
    onAssistantMessageAdded?: (message: Message) => void;
    onAssistantContentUpdated?: (params: {
        messageId: string;
        accumulated: string;
        stage: 'streaming' | 'end';
    }) => void;
    onAssistantReasoningUpdated?: (params: {
        messageId: string;
        accumulated: string;
        stage: 'streaming' | 'end';
    }) => void;
    onToolBlockUpdated?: (params: ToolBlockUpdateCallbackParams) => void;
    onErrorBlockAdded?: (error: string) => void;
    onLoadingChange?: (loading: boolean) => void;
    onCommandRunningChange?: (running: boolean) => void;
    onQueuedMessagesChange?: (messages: QueuedMessage[]) => void;
    onTasksChange?: (tasks: Task[]) => void;
    onSessionIdChange?: (sessionId: string) => void;
    onPermissionModeChange?: (mode: PermissionMode) => void;
    onMcpServersChange?: (servers: McpServerStatus[]) => void;
    onBangMessageAdded?: () => void;
    onBangMessageUpdated?: () => void;
    onBangMessageCompleted?: () => void;
    onNotificationMessageAdded?: (params: {
        taskId: string;
        taskType: string;
        status: string;
        summary: string;
        message?: Message;
    }) => void;
    onPermissionRequest?: (
        requestId: string,
        context: ToolPermissionContext,
    ) => void;
}

// ── StdioAgent ───────────────────────────────────────────────────

export class StdioAgent {
    // ── Cached state (synchronous access) ──
    public sessionId: string | undefined;
    public workingDirectory: string | undefined;
    public latestTotalTokens = 0;
    public permissionMode: PermissionMode | undefined;
    public serverVersion: string | undefined;
    public messages: Message[] = [];
    public queuedMessages: QueuedMessage[] = [];
    public tasks: Task[] = [];

    private client: StdioClient;
    private router: NotificationRouter;
    private callbacks: StdioAgentCallbacks;

    constructor(
        client: StdioClient,
        router: NotificationRouter,
        callbacks: StdioAgentCallbacks,
    ) {
        this.client = client;
        this.router = router;
        this.callbacks = callbacks;
    }

    // ── Lifecycle ─────────────────────────────────────────────────

    async initialize(
        params: InitializeParams,
    ): Promise<InitializeResult> {
        const result = (await this.client.request(
            'initialize',
            params,
        )) as InitializeResult;
        this.sessionId = result.sessionId;
        this.workingDirectory = result.workingDirectory;
        this.permissionMode = result.permissionMode;
        this.latestTotalTokens = result.latestTotalTokens;
        this.serverVersion = result.serverVersion;
        // Register with the router so subsequent notifications are routed here
        this.router.register(this.sessionId, this);
        return result;
    }

    async destroy(): Promise<void> {
        if (this.sessionId) {
            this.router.unregister(this.sessionId);
        }
        try {
            await this.client.request('destroy', undefined, this.sessionId);
        } catch {
            // Process may have already exited; best-effort
        }
        // NOTE: do NOT dispose the shared client — other sessions may still use it
    }

    async restoreSession(sessionId: string): Promise<void> {
        await this.client.request(
            'restoreSession',
            { sessionId },
            this.sessionId,
        );
    }

    async updateConfig(
        params: UpdateConfigParams,
    ): Promise<{ sessionId: string }> {
        const oldSessionId = this.sessionId;
        const result = (await this.client.request(
            'updateConfig',
            params,
            this.sessionId,
        )) as { sessionId: string };
        // If sessionId changed, re-register with the router
        if (oldSessionId && result.sessionId !== oldSessionId) {
            this.router.unregister(oldSessionId);
            this.sessionId = result.sessionId;
            this.router.register(this.sessionId, this);
        }
        return result;
    }

    // ── Messages ──────────────────────────────────────────────────

    async sendMessage(
        text: string,
        images?: Array<{ path: string; mimeType: string }>,
        force?: boolean,
    ): Promise<void> {
        await this.client.request(
            'sendMessage',
            { text, images, force },
            this.sessionId,
        );
    }

    async bang(command: string): Promise<void> {
        await this.client.request('bang', { command }, this.sessionId);
    }

    async abortMessage(): Promise<void> {
        await this.client.request('abortMessage', undefined, this.sessionId);
    }

    async clearMessages(): Promise<void> {
        await this.client.request('clearMessages', undefined, this.sessionId);
    }

    async rewindToMessage(
        messageId: string,
    ): Promise<{ inputContent: string }> {
        return (await this.client.request(
            'rewindToMessage',
            { messageId },
            this.sessionId,
        )) as { inputContent: string };
    }

    async removeQueuedMessage(index: number): Promise<void> {
        await this.client.request(
            'deleteQueuedMessage',
            { index },
            this.sessionId,
        );
    }

    async updateQueuedMessageById(
        id: string,
        patch: {
            content?: string;
            images?: Array<{ path: string; mimeType: string }>;
            type?: 'message' | 'bang';
        },
    ): Promise<boolean> {
        const result = (await this.client.request(
            'updateQueuedMessage',
            { id, text: patch.content, images: patch.images },
            this.sessionId,
        )) as { ok: boolean };
        return result?.ok ?? false;
    }

    async removeQueuedMessageById(id: string): Promise<void> {
        await this.client.request(
            'deleteQueuedMessageById',
            { id },
            this.sessionId,
        );
    }

    async getFullMessageThread(): Promise<{
        messages: Message[];
        sessionIds: string[];
    }> {
        return (await this.client.request(
            'getFullMessageThread',
            undefined,
            this.sessionId,
        )) as { messages: Message[]; sessionIds: string[] };
    }

    // ── Permissions ───────────────────────────────────────────────

    async setPermissionMode(mode: PermissionMode): Promise<void> {
        await this.client.request(
            'setPermissionMode',
            { mode },
            this.sessionId,
        );
    }

    /** Returns cached permission mode (updated via notification). */
    getPermissionMode(): PermissionMode | undefined {
        return this.permissionMode;
    }

    sendPermissionResponse(
        requestId: string,
        decision: PermissionDecision,
    ): void {
        this.client.notify(
            'permissionResponse',
            { requestId, decision },
            this.sessionId,
        );
    }

    // ── MCP ───────────────────────────────────────────────────────

    async getMcpServers(): Promise<McpServerStatus[]> {
        const result = (await this.client.request(
            'getMcpServers',
            undefined,
            this.sessionId,
        )) as { servers: McpServerStatus[] };
        return result.servers;
    }

    async connectMcpServer(serverName: string): Promise<boolean> {
        const result = (await this.client.request(
            'connectMcpServer',
            { serverName },
            this.sessionId,
        )) as { success: boolean };
        return result.success;
    }

    async disconnectMcpServer(serverName: string): Promise<boolean> {
        const result = (await this.client.request(
            'disconnectMcpServer',
            { serverName },
            this.sessionId,
        )) as { success: boolean };
        return result.success;
    }

    // ── Commands ──────────────────────────────────────────────────

    async getSlashCommands(): Promise<SlashCommand[]> {
        const result = (await this.client.request(
            'getSlashCommands',
            undefined,
            this.sessionId,
        )) as { commands: SlashCommand[] };
        return result.commands;
    }

    // ── Notification dispatch (called by NotificationRouter) ──────

    handleNotification(method: string, params: unknown): void {
        switch (method) {
            case 'messagesChange': {
                const p = params as { messages: Message[] };
                this.messages = p.messages;
                this.callbacks.onMessagesChange?.(this.messages);
                break;
            }
            case 'userMessageAdded': {
                const p = params as { message: Message };
                if (p.message) this.callbacks.onUserMessageAdded?.(p.message);
                break;
            }
            case 'assistantMessageAdded': {
                const p = params as { message: Message };
                if (p.message)
                    this.callbacks.onAssistantMessageAdded?.(p.message);
                break;
            }
            case 'assistantContentUpdated':
                this.callbacks.onAssistantContentUpdated?.(
                    params as {
                        messageId: string;
                        accumulated: string;
                        stage: 'streaming' | 'end';
                    },
                );
                break;
            case 'assistantReasoningUpdated':
                this.callbacks.onAssistantReasoningUpdated?.(
                    params as {
                        messageId: string;
                        accumulated: string;
                        stage: 'streaming' | 'end';
                    },
                );
                break;
            case 'toolBlockUpdated':
                this.callbacks.onToolBlockUpdated?.(
                    params as ToolBlockUpdateCallbackParams,
                );
                break;
            case 'errorBlockAdded': {
                const p = params as { error: string };
                this.callbacks.onErrorBlockAdded?.(p.error);
                break;
            }
            case 'loadingChange': {
                const p = params as {
                    loading: boolean;
                    latestTotalTokens: number;
                };
                if (p.latestTotalTokens !== undefined) {
                    this.latestTotalTokens = p.latestTotalTokens;
                }
                this.callbacks.onLoadingChange?.(p.loading);
                break;
            }
            case 'commandRunningChange': {
                const p = params as { running: boolean };
                this.callbacks.onCommandRunningChange?.(p.running);
                break;
            }
            case 'queuedMessagesChange': {
                const p = params as { messages: QueuedMessage[] };
                this.queuedMessages = p.messages;
                this.callbacks.onQueuedMessagesChange?.(p.messages);
                break;
            }
            case 'tasksChange': {
                const p = params as { tasks: Task[] };
                this.tasks = p.tasks;
                this.callbacks.onTasksChange?.(p.tasks);
                break;
            }
            case 'sessionIdChange': {
                const p = params as { sessionId: string };
                this.sessionId = p.sessionId;
                this.callbacks.onSessionIdChange?.(p.sessionId);
                break;
            }
            case 'permissionModeChange': {
                const p = params as { mode: PermissionMode };
                this.permissionMode = p.mode;
                this.callbacks.onPermissionModeChange?.(p.mode);
                break;
            }
            case 'mcpServersChange': {
                const p = params as { servers: McpServerStatus[] };
                this.callbacks.onMcpServersChange?.(p.servers);
                break;
            }
            case 'bangMessageAdded':
                this.callbacks.onBangMessageAdded?.();
                break;
            case 'bangMessageUpdated':
                this.callbacks.onBangMessageUpdated?.();
                break;
            case 'bangMessageCompleted':
                this.callbacks.onBangMessageCompleted?.();
                break;
            case 'notificationMessageAdded':
                this.callbacks.onNotificationMessageAdded?.(
                    params as {
                        taskId: string;
                        taskType: string;
                        status: string;
                        summary: string;
                        message?: Message;
                    },
                );
                break;
            case 'permissionRequest': {
                const p = params as {
                    requestId: string;
                    context: ToolPermissionContext;
                };
                this.callbacks.onPermissionRequest?.(p.requestId, p.context);
                break;
            }
        }
    }
}
